### PR TITLE
[v7] feat: dump_yaml

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/data_classes/_base.py [hunk 1]
```diff
@@ -140,6 +141,15 @@ def dump(self, camel_case: bool = True) -> dict[str, Any]:
         """
         return basic_instance_dump(self, camel_case=camel_case)
 
+    def dump_yaml(self, filepath: str | Path) -> None:
```
Comment (erlendvollset) [inline cognite/client/data_classes/_base.py:144]: I don't find it very intuitive that .dump() returns a dict while .dump_yaml() writes to a file. Let's make it `dump_yaml_to_file()`?

File: cognite/client/data_classes/_base.py [hunk 2]
```diff
@@ -140,6 +141,15 @@ def dump(self, camel_case: bool = True) -> dict[str, Any]:
...
+        """Dump the instance into a YAML file.
+
+        Args:
+            filepath (str | Path): Path to the YAML file.
+        """
+        yaml = local_import("yaml")
+        Path(filepath).write_text(yaml.dump(self.dump(camel_case=True), sort_keys=False))
```
Comment (erlendvollset) [inline cognite/client/data_classes/_base.py:151]: I assume this takes care of closing the file handle after writing?

File: cognite/client/data_classes/_base.py (hunk 1 above)
Comment (gregertw) [reply cognite/client/data_classes/_base.py:144]: Agreed. The expected output from dump_yaml() would be str?

File: cognite/client/data_classes/_base.py [hunk 3]
```diff
@@ -297,6 +307,15 @@ def dump(self, camel_case: bool = True) -> list[dict[str, Any]]:
         """
         return [resource.dump(camel_case) for resource in self.data]
 
+    def dump_yaml(self, filepath: str | Path) -> None:
```
Comment (gregertw) [inline cognite/client/data_classes/_base.py:310]: An alternative could be to send in a stream handler like io.TextIOBase? So do: with open as f: dump_yaml(f)

**Review by gregertw — COMMENTED** [review]
Have an immediate use for writing yaml to a file, but would be useful to get an in-memory representation as well.

File: cognite/client/data_classes/_base.py (hunk 1 above)
Comment (erlendvollset) [reply cognite/client/data_classes/_base.py:144]: Yes I'd say so

File: cognite/client/data_classes/_base.py (hunk 3 above)
Comment (erlendvollset) [reply cognite/client/data_classes/_base.py:310]: (or both)

**Comment by erlendvollset** [comment]
Let's also add a test

**Comment by haakonvt** [comment]
I think we should call the function `to_yaml` and have it return a string. It can then optionally accept a file handle, and output to that instead.

File: CHANGELOG.md [hunk 4]
```diff
@@ -86,6 +86,7 @@ with no easy way to add a prefix. Also, it no longer expands metadata by default
 - In transformations, `NonceCredentials` was missing `load` method.
 - In transformations, `TransformationBlockedInfo` was missing `.dump` method
 - `capabilities` in `cognite.client.data_classes` with data classes for all CDF capabilities.
+- All `CogniteResource` and `CogniteResourcelist` objects have  `.dump_yaml` methdos, for example, `my_asset_list.dump_yaml("my_assets.yaml")`.
```
Comment (haakonvt) [inline CHANGELOG.md:89]:
```suggestion
- All `CogniteResource` and `CogniteResourcelist` objects have  `.dump_yaml` methods, for example, `my_asset_list.dump_yaml("my_assets.yaml")`.  
```

File: cognite/client/data_classes/_base.py (hunk 1 above)
Comment (doctrino) [reply cognite/client/data_classes/_base.py:144]: Agree, and I also think Peter suggestion is a better way, just return the string.

File: cognite/client/data_classes/_base.py (hunk 2 above)
Comment (doctrino) [reply cognite/client/data_classes/_base.py:151]: Yes it does.

**Comment by doctrino** [comment]
> Let's also add a test

Modified the existing `test_yaml_serialize` to test this on all classes.

File: cognite/client/data_classes/_base.py [hunk 5]
```diff
@@ -297,6 +306,15 @@ def dump(self, camel_case: bool = True) -> list[dict[str, Any]]:
         """
         return [resource.dump(camel_case) for resource in self.data]
 
+    def dump_yaml(self) -> str:
+        """Dump the instance into a YAML formatted string.
+
+        Returns:
+            str: A YAML formatted string representing the instance.
```
Comment (doctrino) [inline cognite/client/data_classes/_base.py:313]:
```suggestion
            str: A YAML formatted string representing the instances.
```

File: cognite/client/data_classes/_base.py [hunk 6]
```diff
@@ -297,6 +306,15 @@ def dump(self, camel_case: bool = True) -> list[dict[str, Any]]:
         """
         return [resource.dump(camel_case) for resource in self.data]
 
+    def dump_yaml(self) -> str:
+        """Dump the instance into a YAML formatted string.
```
Comment (doctrino) [inline cognite/client/data_classes/_base.py:310]:
```suggestion
        """Dump the instances into a YAML formatted string.
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -86,6 +86,7 @@ with no easy way to add a prefix. Also, it no longer expands metadata by default
 - In transformations, `NonceCredentials` was missing `load` method.
 - In transformations, `TransformationBlockedInfo` was missing `.dump` method
 - `capabilities` in `cognite.client.data_classes` with data classes for all CDF capabilities.
+- All `CogniteResource` and `CogniteResourcelist` objects have  `.dump_yaml` methods, for example, `my_asset_list.dump_yaml("my_assets.yaml")`.
```
Comment (haakonvt) [inline CHANGELOG.md:89]:
```suggestion
- All `CogniteResource` and `CogniteResourcelist` objects have  `.dump_yaml` methods, for example, `my_asset_list.dump_yaml()`,
  returning a yaml-formatted string.
```

**Review by haakonvt — APPROVED** [review]
Lgtm!

**Review by erlendvollset — APPROVED** [review]
Lgtm

**Pull request merged**